src/core/address_payload_unification_design.md

```diff
@@ -0,0 +1,556 @@
+class GithubPluginHandler
+implements PluginHandler<GithubNodeReference, GithubNodePayload<any>> {
+  createReference(reference) {
+    const address = reference.address();
+    if (address.pluginName !== "sourcecred/github-beta") {
+      throw new Error("pluginName: " + address.pluginName);
+    }
+    const type: GithubNodeType = ((address.type: string): any);
+    switch (type) {
+      case "PULL_REQUEST":
+        return new PullRequestReference(reference);
+      case "PULL_REQUEST_COMMENT":
+        return new PullRequestCommentReference(reference);
+      default:
+        // eslint-disable-next-line no-unused-expressions
+        (type: empty);
+        throw new Error("type: " + type);
+    }
+  }
+
+  fromJSON(json: any) {
+    const type: GithubNodeType = ((json.type: string): any);
+    switch (type) {
+      case "PULL_REQUEST":
+        return new PullRequestPayload(json);
+      case "PULL_REQUEST_COMMENT":
+        return new PullRequestCommentPayload(json);
+      default:
+        // eslint-disable-next-line no-unused-expressions
+        (type: empty);
+        throw new Error("type: " + type);
+    }
+  }
+}
+```
+
+## The `Graph` class
+
+A graph is associated with a fixed set of plugins:
+
+```javascript
+type Plugins = {[pluginName: string]: PluginHandler<any, any>};
+```
+
+The graph learns about these at construction time, either via the
+constructor or the `static fromJSON` method.
+
+Graphs, of course, also need to refer to edges. In this proposal, we
+leave edges structurally unchanged, but remove their type parameters:
+
+```javascript
+type Edge = {|
+  +address: Address,
+  +src: Address,
+  +dst: Address,
+  +payload: any,
+|};
+```
+
+The `Graph` class then looks something like this:
+
+```javascript
+declare class Graph /* no type parameters! */ {
+  constructor(plugins: Plugins): void;
+
+  copy(): Graph;
+  equals(that: Graph): boolean;
+
+  toJSON(): GraphJson;
+  static fromJSON(json: GraphJson, plugins: Plugins): Graph;
+
+  addNode(np: NodePayload): this;
+  addEdge(edge: Edge): this;
+
+  removeNode(address: Address): this;
+  removeEdge(address: Address): this;
+
+  node(address: Address): ?Node<any, any>;
+  edge(address: Address): ?Edge;
+  nodeReference(address: Address): NodeReference /* see commentary below */;
+
+  nodes(filter?: {|+type?: string|}): Iterator<Node<any, any>>;
+  edges(filter?: {|+type?: string|}): Iterator<Edge>;
+
+  static mergeConservative(Iterator<Graph>): Graph;
+
+  // Some implementation details (but not all of them)…
+  _nodes: {|+address: Address, +value: Node<any, any> | void|}[];
+  _indexVersion: number;
+  _internalNeighbors(
+    number: number,
+    options?: NeighborhoodOptions
+  ): Iterator<NodeReference>;
+  _addressToIndex(address: Address): ?number;
+}
+type GraphJson = {/* elided */};
+```
+
+I’ll draw attention to a few things in this proposal, which I am open to
+changing.
+
+  - APIs provide iterators by default. Conversion with `Array.from` is
```

Comment:
Seems reasonable, I don't 100% love it because it adds some extra boilerplate, but I do like defaulting to giving us more expressive / flexible semantics.